src/com/linkedin/parseq/internal/Continuations.java

```diff
@@ -44,51 +42,54 @@ public void submit(final Runnable action) {
   }
 
   private static final class Continuation {
-    // contains actions scheduled in the current recurrence level
-    private Deque<Runnable> _active;
+    // contains sibling actions in reverse order of submission
+    // sibling actions are actions submitted by the same parent action
+    private final Deque<Runnable> _siblingActions = new ArrayDeque<>();
 
-    // variable to cache empty deque instance
-    private Deque<Runnable> _inactive;
-
-    // contains actions scheduled for execution
-    private Deque<Runnable> _scheduled;
+    /** contains actions for execution in post-order
+     *  actions with larger depth at the top;
+     *  for sibling actions with the same depth, the first submitted is at the top
+     */
+    private final Deque<Runnable> _postOrderExecutionStack = new ArrayDeque<>();
 
     private void submit(final Runnable action) {
-      if (_active == null) {
+      if (_postOrderExecutionStack.size() == 0) {
```

Comment:
It would be even cleaner to have a nullable variable _activeAction that is set in submit() and nulled out in the finally block in the loop(). This way condition here can change to:
if (_activeAction == null)

and loop can further simplify to just do _postOrderExecutionStack.pop() instead of peek/pop combination.